contracts/marketplace/src/lib.rs

```diff
@@ -0,0 +1,215 @@
+        buyer.require_auth();
+        let storage = env.storage().persistent();
+        let token = storage.get(&DataKey::Token).unwrap();
+        let mut assets: Map<Address, Asset> = storage.get(&DataKey::Assets).unwrap();
+        let Asset {
+            owner: seller,
+            price: set_price,
+            listed,
+        } = assets.get(asset.clone()).unwrap();
+        if !listed {
+            panic_with_error!(&env, Error::AssetNotListed);
+        }
+        if price != set_price {
+            panic_with_error!(&env, Error::InvalidAssetPrice);
+        }
+        let token = token::Client::new(&env, &token);
+        if token.balance(&buyer) < price {
+            panic_with_error!(&env, Error::BalanceTooLow);
+        }
+        token.transfer(&buyer, &seller, &price);
+        assets.set(
+            asset,
+            Asset {
+                owner: buyer,
+                price,
+                listed: false,
+            },
+        );
+        storage.set(&DataKey::Assets, &assets);
+    }
+
+    /// Permit sellers to update the price of a listed asset,
+    /// ensuring they provide the correct seller and asset addresses, as well as the old and new prices.
+    pub fn update_listing(
+        env: Env,
+        seller: Address,
+        asset: Address,
+        old_price: i128,
+        new_price: i128,
+        listed: bool,
+    ) {
+        if new_price <= 0 {
+            panic_with_error!(&env, Error::InvalidAssetPrice);
+        }
+        seller.require_auth();
+        let storage = env.storage().persistent();
+        let mut assets: Map<Address, Asset> = storage.get(&DataKey::Assets).unwrap();
+        let Asset {
+            owner: set_seller,
+            price: set_price,
+            ..
+        } = assets.get(asset.clone()).unwrap();
+        assert_eq!(set_seller, seller);
+
+        if old_price != set_price {
+            panic_with_error!(&env, Error::InvalidAssetPrice);
+        }
+        assets.set(
+            asset,
+            Asset {
+                owner: seller,
+                price: new_price,
+                listed,
+            },
+        );
+        storage.set(&DataKey::Assets, &assets);
+    }
+
+    /// Allow sellers to pause a listing by specifying their address, the asset's address,
+    /// and the price at which it was listed.
+    pub fn pause_listing(env: Env, seller: Address, asset: Address, price: i128) {
+        if price <= 0 {
+            panic_with_error!(&env, Error::InvalidAssetPrice);
+        }
+        seller.require_auth();
+        let storage = env.storage().persistent();
+        let mut assets: Map<Address, Asset> = storage.get(&DataKey::Assets).unwrap();
+        let Asset {
+            owner: set_seller,
+            price: set_price,
+            ..
+        } = assets.get(asset.clone()).unwrap();
+        assert_eq!(set_seller, seller);
+        if price != set_price {
+            panic_with_error!(&env, Error::InvalidAssetPrice);
+        }
```

Comment:
Also here, once `assert_eq` is used and then `if` + `panic_with_error `

Reply:
One just panics if you are not the owner.
The other returns an error that your input is wrong.
All these just do the same thing only that `panic_with_error` adds some specific error.

Reply:
Maybe better add an error saying user is not an owner? It would be more clear what is wrong.